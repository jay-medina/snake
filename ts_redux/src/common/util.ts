import { GridItem, Snake, AppState, Apple } from './types';

const isAtPosition = (gridItem1: GridItem) => (gridItem2: GridItem) => {
  return gridItem1.row === gridItem2.row && gridItem1.col === gridItem2.col;
};

const randomNumber = (num: number) => {
  return Math.floor(Math.random() * num);
};

const randomGridItem = (dimensions: AppState['dimensions']): GridItem => {
  return {
    row: randomNumber(dimensions.rows),
    col: randomNumber(dimensions.cols),
  };
};

export const isTheApple = isAtPosition;

export const isSnakeAtPosition = (snake: Snake['body']) => (gridItem: GridItem) =>
  !!snake.find(isAtPosition(gridItem));

export function findNewApplePosition(appState: AppState): Apple {
  const { snake, dimensions } = appState;

  let newApple = randomGridItem(dimensions);

  while (isSnakeAtPosition(snake.body)(newApple)) {
    newApple = randomGridItem(dimensions);
  }

  return newApple;
}

export function isSnakeDead(appState: AppState) {
  const { snake } = appState;
  return isSnakeInWall(appState) || isSnakeAtItself(snake.body);
}

function isSnakeInWall({ snake, dimensions }: AppState) {
  const [{ row, col }] = snake.body;
  const { rows, cols } = dimensions;

  return row < 0 || row >= rows || col < 0 || col >= cols;
}

function isSnakeAtItself(snake: Snake['body']) {
  const [head, ...rest] = snake;

  return isSnakeAtPosition(rest)(head);
}
