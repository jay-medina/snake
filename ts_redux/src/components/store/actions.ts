// import { GameThunkAction } from './types';
// import { AppState } from './util';

interface GameAction {
  type: 'START_GAME';
}

export type AppAction = GameAction;

export const startGame = (): GameAction => ({
  type: 'START_GAME',
});

// export function randomizeApple(appState: AppState) {

// }

// export const startGameThunk = (): GameThunkAction => (dispatch) => {
//   // randomize the apple position
//   // start the game
// //  return dispatch(startGame);
// };
