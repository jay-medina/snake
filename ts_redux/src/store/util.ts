export enum GameState {
  Start,
  Run,
  GameOver,
}

export interface AppState {
  gameState: GameState;
  apple: Apple;
  snake: Snake;
  dimensions: {
    rows: number;
    cols: number;
  };
}

export interface GridItem {
  row: number;
  col: number;
}

export type Apple = GridItem;

export type Snake = GridItem[];
