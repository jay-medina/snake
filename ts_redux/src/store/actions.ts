import { GameThunkAction, Apple, TimeStamp, Direction } from '../common/types';
import { findNewApplePosition, isSnakeAtPosition } from '../common/util';

interface GameAction {
  type: 'START_GAME';
}

interface AppleAction {
  type: 'UPDATE_APPLE';
  payload: {
    apple: Apple;
  };
}

interface ScoreAction {
  type: 'INCREASE_SCORE';
}

export interface TickTimeAction {
  type: 'TICK_TIME';
  payload: {
    timestamp: TimeStamp;
  };
}

export interface SnakeDirectionAction {
  type: 'UPDATE_SNAKE_DIRECTION';
  payload: {
    direction: Direction;
  };
}

export interface SnakeSizeAction {
  type: 'UPDATE_SNAKE_SIZE';
}

export type SnakeAction = SnakeDirectionAction | SnakeSizeAction;

export type AppAction = GameAction | AppleAction | TickTimeAction | SnakeAction | ScoreAction;

export const startGame = (): GameAction => ({
  type: 'START_GAME',
});

export const newApplePosition = (position: Apple): AppleAction => ({
  type: 'UPDATE_APPLE',
  payload: {
    apple: position,
  },
});

export const tickForward = (timestamp: TimeStamp): TickTimeAction => ({
  type: 'TICK_TIME',
  payload: {
    timestamp,
  },
});

export const growSnake = (): SnakeSizeAction => ({
  type: 'UPDATE_SNAKE_SIZE',
});

export const updateScore = (): ScoreAction => ({
  type: 'INCREASE_SCORE',
});

export const newAppleThunk = (): GameThunkAction => (dispatch, getState) => {
  const appState = getState();

  const newApple = findNewApplePosition(appState);
  return dispatch(newApplePosition(newApple));
};

export const tickForwardThunk = (timestamp: TimeStamp): GameThunkAction => (dispatch, getState) => {
  dispatch(tickForward(timestamp));

  const appState = getState();
  const { snake, apple } = appState;

  if (isSnakeAtPosition(snake.body)(apple)) {
    dispatch(growSnake());
    dispatch(newAppleThunk());
    dispatch(updateScore());
  }
};

export const startGameThunk = (): GameThunkAction => (dispatch) => {
  dispatch(startGame());
  dispatch(newAppleThunk());
};

export const updateSnakeDirection = (direction: Direction): SnakeDirectionAction => ({
  type: 'UPDATE_SNAKE_DIRECTION',
  payload: {
    direction,
  },
});
