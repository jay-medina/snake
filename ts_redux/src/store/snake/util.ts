import { Snake, TimeStamp, Direction } from '../../common/types'

export function isSnakeAbleToMove(snake: Snake, timestamp: TimeStamp) {
  return (
    snake.lastTimestamp === 0 ||
    timestamp - snake.lastTimestamp >= snake.incrementTimer
  )
}

export function moveSnake(snake: Snake): Snake['body'] {
  const { body, direction } = snake
  const [head] = body

  if (direction === 'right') {
    return [{ row: head.row, col: head.col + 1 }, ...body.slice(0, body.length - 1)]
  } else if (direction === 'left') {
    return [{ row: head.row, col: head.col - 1 }, ...body.slice(0, body.length - 1)]
  } else if (direction === 'up') {
    return [{ row: head.row - 1, col: head.col }, ...body.slice(0, body.length - 1)]
  } else {
    return [{ row: head.row + 1, col: head.col }, ...body.slice(0, body.length - 1)]
  }
}

export const isOppositeDirection = (direction: Direction, newDirection: Direction) => {
  return (
    (direction === 'left' && newDirection === 'right') ||
    (direction === 'right' && newDirection === 'left') ||
    (direction === 'up' && newDirection === 'down') ||
    (direction === 'down' && newDirection === 'up')
  )
}
