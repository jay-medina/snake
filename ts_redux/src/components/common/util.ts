import { GridItem, Snake } from 'store/util';

const isAtPosition = (gridItem1: GridItem) => (gridItem2: GridItem) => {
  return gridItem1.row === gridItem2.row && gridItem1.col === gridItem2.col;
};

export const isSnakeAtPosition = (snake: Snake) => (gridItem: GridItem) =>
  !!snake.find(isAtPosition(gridItem));

export const isTheApple = isAtPosition;
