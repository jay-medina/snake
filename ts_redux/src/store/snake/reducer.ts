import { Snake } from '../../common/types';
import { AppAction } from '../actions';
import { isSnakeAbleToMove, moveSnake, isOppositeDirection } from './util';

function updateDirection(snake: Snake): Snake {
  if (snake.newDirection) {
    return {
      ...snake,
      direction: snake.newDirection,
      newDirection: undefined,
    };
  }

  return snake;
}

// TODO determine if snake is ded
export const snakeReducer = (snake: Snake, action: AppAction): Snake => {
  if (action.type === 'TICK_TIME') {
    const { payload } = action;
    const newSnake = updateDirection(snake);

    if (isSnakeAbleToMove(newSnake, payload.timestamp)) {
      return {
        ...newSnake,
        body: moveSnake(newSnake),
        lastTimestamp: payload.timestamp,
      };
    }
  }

  if (action.type === 'UPDATE_SNAKE_DIRECTION') {
    const { payload } = action;

    if (!isOppositeDirection(snake.direction, payload.direction)) {
      return {
        ...snake,
        newDirection: payload.direction,
      };
    }
  }

  if (action.type === 'UPDATE_SNAKE_SIZE') {
    const newSnake = updateDirection(snake);
    const { body } = newSnake;
    const lastPosition = body[body.length - 1];

    return {
      ...newSnake,
      body: [...body, lastPosition],
    };
  }

  return snake;
};
