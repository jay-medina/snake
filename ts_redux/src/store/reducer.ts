import { AppAction } from './actions';
import { GameState, Apple, AppState, Score } from '../common/types';
import { snakeReducer } from './snake/reducer';

const initialState: AppState = {
  score: {
    current: 0,
    high: 100,
  },
  dimensions: {
    cols: 25,
    rows: 25,
  },
  apple: {
    row: 2,
    col: 2,
  },
  snake: {
    body: [{ row: 5, col: 3 }, { row: 5, col: 2 }, { row: 5, col: 1 }],
    incrementTimer: 200,
    lastTimestamp: 0,
    direction: 'right',
  },
  gameState: 'Start',
};

const gameState = (state: GameState, action: AppAction): GameState => {
  if (action.type === 'START_GAME') {
    return 'Run';
  }

  return state;
};

const apple = (state: Apple, action: AppAction): Apple => {
  if (action.type === 'UPDATE_APPLE') {
    return action.payload.apple;
  }

  return state;
};

const score = (state: Score, action: AppAction): Score => {
  if (action.type === 'INCREASE_SCORE') {
    const { current, high } = state;
    const newCurrent = current + 10;

    return {
      current: newCurrent,
      high: high > newCurrent ? high : newCurrent,
    };
  }

  return state;
};

export const app = (state: AppState = initialState, action: AppAction): AppState => {
    return {
      ...state,
      snake: snakeReducer(state.snake, action),
      apple: apple(state.apple, action),
      gameState: gameState(state.gameState, action),
      score: score(state.score, action),
    };
};
