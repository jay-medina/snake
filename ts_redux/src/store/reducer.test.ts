import { app } from './reducer';
import { GameState } from './util';
import { startGame } from './actions';

describe('game', () => {
  it('initializes to start', () => {
    const action: any = {};

    const nextState = app(undefined, action);
    expect(nextState.gameState).toEqual(GameState.Start);
  });

  describe('when the action is to start the game', () => {
    it('sets the game to run', () => {
      const action = startGame();

      const nextState = app(undefined, action);
      expect(nextState.gameState).toEqual(GameState.Run);
    });
  });
});

describe('apple', () => {
  it('initializes the apple', () => {
    const action: any = {};

    const nextState = app(undefined, action);
    expect(nextState.apple).toEqual({
      row: 4,
      col: 4,
    });
  });
});
