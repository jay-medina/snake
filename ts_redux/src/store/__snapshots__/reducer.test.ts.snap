// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app creates the app state 1`] = `
Object {
  "apple": Object {
    "col": 2,
    "row": 2,
  },
  "dimensions": Object {
    "cols": 25,
    "rows": 25,
  },
  "gameState": "Start",
  "score": Object {
    "current": 0,
    "high": 100,
  },
  "snake": Object {
    "body": Array [
      Object {
        "col": 3,
        "row": 5,
      },
      Object {
        "col": 2,
        "row": 5,
      },
      Object {
        "col": 1,
        "row": 5,
      },
    ],
    "direction": "right",
    "incrementTimer": 200,
    "lastTimestamp": 0,
  },
}
`;
